docs/source/final_attrs.rst

```diff
@@ -0,0 +1,267 @@
+used in one of these forms:
+
+
+* You can provide an explicit type using the syntax ``Final[<type>]``. Example:
+
+  .. code-block:: python
+
+     ID: Final[float] = 1
+
+* You can omit the type: ``ID: Final = 1``. Note that unlike for generic
+  classes this is *not* the same as ``Final[Any]``. Here mypy will infer
+  type ``int``.
+
+* In stub files you can omit the right hand side and just write
+  ``ID: Final[float]``.
+
+* Finally, you can define ``self.id: Final = 1`` (also with a type argument),
+  but this is allowed *only* in ``__init__`` methods.
+
+Definition rules
+****************
+
+The are two rules that should be always followed when defining a final name:
+
+* There can be *at most one* final declaration per module or class for
+  a given attribute:
+
+  .. code-block:: python
+
+     from typing_extensions import Final
+
+     ID: Final = 1
+     ID: Final = 2  # Error: "ID" already declared as final
+
+     class SomeCls:
+         id: Final = 1
+         def __init__(self, x: int) -> None:
+             self.id: Final = x  # Error: "id" already declared in class body
+
+  Note that mypy has a single namespace for a class. So there can't be two
+  class-level and instance-level constants with the same name.
+
+* There must be *exactly one* assignment to a final attribute:
+
+  .. code-block:: python
+
+     ID = 1
+     ID: Final = 2  # Error!
+
+     class SomeCls:
+         ID = 1
+         ID: Final = 2  # Error!
+
+* A final attribute declared in class body without an initializer must
+  be initialized in the ``__init__`` method (you can skip the initializer
+  in stub files):
+
+  .. code-block:: python
+
+     class SomeCls:
+         x: Final[int]
+         y: Final[int]  # Error: final attribute without an initializer
+         def __init__(self) -> None:
+             self.x = 1  # Good
+
+* ``Final`` can be only used as an outermost type in assignments or variable
+  annotations. using it in any other position is an error. In particular,
+  ``Final`` can't be used in annotations for function arguments:
+
+  .. code-block:: python
+
+     x: List[Final[int]] = []  # Error!
+     def fun(x: Final[List[int]]) ->  None:  # Error!
+         ...
+
+* ``Final`` and ``ClassVar`` should not be used together. Mypy will infer
+  the scope of a final declaration automatically depending on whether it was
+  initialized in the class body or in ``__init__``.
+
+Using final attributes
+**********************
+
+As a result of a final declaration mypy strives to provide the
+two following guarantees:
+
+* A final attribute can't be re-assigned (or otherwise re-defined), both
+  internally and externally:
+
+  .. code-block:: python
+
+     # file mod.py
+     from typing_extensions import Final
+
+     ID: Final = 1
+
+     class SomeCls:
+         ID: Final = 1
+
+         def meth(self) -> None:
+             self.ID = 2  # Error: can't assign to final attribute
```

Comment:
Again I think it would be less confusing if the module variable and the class variable had different names.